CasperLabs/casper-node
# NDRS-{866|867|868}: Joining node synchronizes with the upgraded chain.

## Description
While executing linear chain blocks (from Genesis up until fully synchronized), node will check on every switch block if it matches the scheduled upgrade. If it does, then it will persist its current state and shut down. After the restart, it will read its last state and continue where it left off. The process repeats until fully synchronized. Once the node is synchronized it will clean the snapshot state of the syncing component.

## JIRA tickets:
https://casperlabs.atlassian.net/browse/NDRS-866
https://casperlabs.atlassian.net/browse/NDRS-867
https://casperlabs.atlassian.net/browse/NDRS-868

## How to test:
1. Install `nctl`.
2. Build packages: `nctl-compile`.
3. Start the network: `nctl-start` (starts 5 nodes by default).
4. Wait for a couple of eras: `nctl-view-chain-era`
5. In the root directory of idle node (when using defaults, `node-6` is idle), create a new directory `2_0_0`.
6. Copy the chainspec to it (`cp config/chainspec.toml 2_0_0`) and modify it to reflect a scheduled upgrade: `protocol_version` needs to be set to the same value as the folder (`2_0_0` in this case) and `protocol.activation_point.era_id` to some non-genesis era that is lower than the current era of the network.
7. Take a note of trusted hash (`nctl-view-chain-lfb node=1`) and put it in the node config of the new node (`config/node-config.toml` under `node.trusted_hash`). 
8. Start the new node: `nctl-start node=6`.
9. Observe that the new node synchronizes until the end of an era that is 1 lower than the value set for `protocol.activation_point.era_id` in the `2_0_0` chainspec. `node-6` should shut down.

Since we also want to test that the node resumes after restart, we need to simulate that the upgrade happened and we're ready to continue. To do that, we will remove the `2_0_0` directory. `node-6` should pick up where it left and finish synchronization.

10. Set `node.trusted_hash`, in the genesis chainspec of `node-6`, to a trusted hash (can be a new one).
11. Start the node: `nctl-start node=6`.
12. Check that the node reuses previous state: `cat <node-6 log file> | grep "reusing"` should log the pre-shutdown state.
13. Observe that the node finishes synchronization: `tail -f <node-6 log file> | grep "finished joining"`.
14. Check that all nodes are at the same LFB: `nctl-view-chain-lfb`.

### Notes
I also tested that the state is cleared once the node is synchronized. 

To reproduce:

Once the node is synchronized, simply shut it down and restart with a new trusted hash. It won't reuse the previous state and start the joining process from scratch.
